ships/0025-strategy-parameters-enhancements.md

```diff
@@ -0,0 +1,281 @@
+- Allow build users to define values for a string parameter, or an element of an array parameter, using a ConfigMap key reference.
+- Allow build users to define a prefix for the Secret value.
+- Allow build users to define a prefix for a ConfigMap value.
+
+### Non-Goals
+
+## Proposal
+
+### API Changes
+
+We will extend the Build Strategy Parameter type to support the following:
+
+```yaml
+spec:
+  parameters:
+    - name: build-args
+      description: The values for the ARGs in the Dockerfile.
+      type: array
+      defaults: []
+```
+
+The `type` property is new and allows either `string` or `array` with `string` being the default.
+
+The `defaults` parameter is new and is a string array. No possibility here to reference Secret or ConfigMap values.
+
+The Build and BuildRun parameterValue will be extended to support the following:
+
+```yaml
+spec:
+  paramValues:
+    - name: single-secret-value
+      secretValue:
+        name: secret-name
+        key: secret-key
+    - name: single-configmap-value
+      configMapValue:
+        name: configmap-name
+        key: configmap-key
+    - name: simple-list
+      values:
+        - value: first-item
+        - value: second-item
+    - name: build-args
+      values:
+        - value: KEY1=myarg1
+        - secretValue:
+            name: build-arg-secret
+            key: value
+            format: "KEY2=${SECRET_VALUE}" 
+        - configMapValue:
+            name: build-configuration
+            key: base-image-digest
+            format: "BASE_IMAGE_DIGEST=${CONFIGMAP_VALUE}"
+```
+
+The `secretValue` property is new and can be used instead of value to reference a key in a Secret. Optionally, the `format` property (shown above at the end inside an array) can be used to add text around the Secret value. The scenario is clear there: the value for the build arg comes from a Secret, but the key of the build arg also needs to go there.
+
+The `configMapValue` property is also new and works in the same way as the `secretValue`. A `format` is also supported here.
+
+The `values` property is new and allows to specify multiple values. Each value can either be a simple string `value`, or a `secretValue`, or a `configMapValue`.
+
+Validation will be added ...
+
+- to the Build and BuildRun to verify that the value for an array parameter is not a single value, and vice versa.
+- to the Build and BuildRun to verify the existence of the Secret and the referenced key.
+- to the Build and BuildRun to verify the existence of the ConfigMap and the referenced key.
+
+### Backend
+
+We will continue to use Tekton results. This is straight-forward for the array support.
+
+For Secrets, we will do the following:
+
+- For each `secretKeyValue`, we will add an environment variable to those steps that reference the param.
+  - The environment variable name will be `SHP_SECRET_PARAM_{randomSuffix}`
```

Comment:
Assuming my confusion around `secretKeyValue` is not getting in the way, my initial thinking was this environment variable name included a way to correlate to the `paramValue`

in other words
- the `name` of the `paramValue`
- the secret name and secret key

given the multiple params and multiple keys within a secret that are possible

do you agree that should be included?  or was connection / assumption am I missing that makes those clarifications irrelevant?

thanks

Reply:
No, actually I had not planned to a way to provide any means to look at the environment variable name and correlate this to some paramValue. Several reasons:

* A paramValue can be an array and then just one item in there references a secret. I then need to include indices in the env var name as well.
* A paramValue's key is a Kubernetes name, so lower case letters, numbers and hyphens. Environment variable names don't like hyphens.
* Obfuscation. I know it does not give us any additional security at the end, but if the environment variable name has no indication of what it contains, then it is at least a small amount of added security (`SHP_SECRET_PARAM_ABCDE` vs `SHP_SECRET_PARAM_GITHUB_TOKEN` - guess what the first one's value is for and repeat the same for the second one ;-) )

Reply:
yeah end of the day it is an internal shipwright implementation detail, where the consumer deals `$(params.some-var)` in his associated strategy

OK I think I'm good on this one